Pod html oo 20210511

jkeenan opened the pull request:
This pull request is the second of five, the result of which will be that Pod::Html will now internally be object-oriented.

jkeenan committed:
Introduce Testing::record_state_of_cache()

A method to assist in debugging cache problems.

Should assist in resolving https://github.com/Perl/perl5/issues/12271.

Signed-off-by: James E Keenan <jkeenan@cpan.org>

jkeenan committed:
Reduce default arguments for Testing:_prepare_argstable()

Touch-up t/htmlview.t

jkeenan committed:
Start refactoring lib/Pod/Html.pm

Group all the "global" variables into a hashref $globals.  This variable
will be the return value of init_globals() within pod2html(), will be
passed through, and augmented by, parse_command_line() and then used
through the balance of pod2html().

In essence, so far this is just a renaming of variables.

Remove declarations of superseded variables.

Pass $globals to get_cache(), cache_key() and load_cache().

jkeenan committed:
Introduce Pod::Html::Auxiliary

This package will hold helper subroutines used within the main package
or in tests.  They can be placed in a separate module and imported into
Pod::Html because they won't depend on having the globals passed as an
argument.  They will also be potentially independently testable.

Start with html_escape().

Move anchorify(), htmlify() to Auxiliary.pm.  Also _unixify -- now as unixify().

Move relativize_url() to Auxiliary.

Move usage() to Auxiliary.

Move trim_leading_whitespace to Auxiliary.

Move parse_command_line() to Auxiliary.

Keep porting tests happy. Increment $VERSION.  Run: ./perl -Ilib
regen/lib_cleanup.pl

anchorify.t, eol.t:  Correct excessive corrections.

Standardize setting of $VERSION.

jkeenan committed:
Create sub write_file()

We now start placing parts of sub pod2html() into separate subs so that
we reduce the length of pod2html(), making it more readable and
(ultimately) preparing for method calls.

Signed-off-by: James E Keenan <jkeenan@cpan.org>

jkeenan committed:
Remove last hard-tabs from Html.pm

Signed-off-by: James E Keenan <jkeenan@cpan.org>

jkeenan committed:
Create internal sub parse_input_for_podtree()

This clears up a bit of semantic confusion.  We were using one lexically
scoped variable -- my $parser -- to hold two different objects: one for
parsing input, one for writing output.  We can encapsulate the working
of the input parser to make the code more readable.

Signed-off-by: James E Keenan <jkeenan@cpan.org>

jkeenan committed:
Introduce internal sub refine_globals()

Encapsulate! Encapsulate!

Signed-off-by: James E Keenan <jkeenan@cpan.org>

jkeenan committed:
Introduce internal sub set_Title();

Encapsulate more code within pod2html(). The return value is for
convenience.  It's the setting inside $globals that counts.

Signed-off-by: James E Keenan <jkeenan@cpan.org>

jkeenan committed:
Introduce generate_cache() and identify_input()

Further encapsulation of code to improve readability.

Signed-off-by: James E Keenan <jkeenan@cpan.org>

jkeenan committed:
Introduce refine_parser()

Further encapsulation of code internal to pod2html().

At this point we face obstacles:  %Pages; $output.

Signed-off-by: James E Keenan <jkeenan@cpan.org>

jkeenan committed:
Rename one method per suggestion from rjbs

oodler577 commented on ext/Pod-Html/lib/Pod/Html/Auxiliary.pm:
```diff
@@ -0,0 +1,292 @@
+        $opt_outfile,$opt_poderrors,$opt_podpath,$opt_podroot,
+        $opt_quiet,$opt_recurse,$opt_title,$opt_verbose);
+
+    unshift @ARGV, split ' ', $Config{pod2html} if $Config{pod2html};
```
Because modifying `@ARGV` like this may effect things outside this subroutine (unless that's the intent), I would at least localize `@ARGV`, either with `local`, or even:
```
my @_argv = @ARGV ;
unshift @_argv, split( ' ', $Config{pod2html} || '');
my $result = GetOptionsFromArray(
                       \@_argv,
                       'backlink!'  => \$opt_backlink,
                       'cachedir=s' => \$opt_cachedir,
                       'css=s'      => \$opt_css,
                       'flush'      => \$opt_flush,
                       'help'       => \$opt_help,
                       'header!'    => \$opt_header,
                       'htmldir=s'  => \$opt_htmldir,
                       'htmlroot=s' => \$opt_htmlroot,
                       'index!'     => \$opt_index,
                       'infile=s'   => \$opt_infile,
                       'outfile=s'  => \$opt_outfile,
                       'poderrors!' => \$opt_poderrors,
                       'podpath=s'  => \$opt_podpath,
                       'podroot=s'  => \$opt_podroot,
                       'quiet!'     => \$opt_quiet,
                       'recurse!'   => \$opt_recurse,
                       'title=s'    => \$opt_title,
                       'verbose!'   => \$opt_verbose,
);
```

I see some other opportunites to organize this subroutine for clearity and maintenance - would you be willing to accept a PR for it after it gets an initial merge?

oodler577 reviewed (commented):
I noticed only 1 potentially unsafe thing, left a more detailed comment near the line number.

jkeenan commented:
On 5/14/21 12:21 AM, oodler577 wrote:
> ***@***.**** commented on this pull request.
> 
> ------------------------------------------------------------------------
> 
> In ext/Pod-Html/lib/Pod/Html/Auxiliary.pm 
> <https://github.com/jkeenan/perl5/pull/4#discussion_r632271870>:
> 
>> +=head1 SUBROUTINES
> +
> +=head2 C<parse_command_line()>
> +
> +TK
> +
> +=cut
> +
> +sub parse_command_line {
> +    my $globals = shift;
> +    my ($opt_backlink,$opt_cachedir,$opt_css,$opt_flush,$opt_header,
> +        $opt_help,$opt_htmldir,$opt_htmlroot,$opt_index,$opt_infile,
> +        $opt_outfile,$opt_poderrors,$opt_podpath,$opt_podroot,
> +        $opt_quiet,$opt_recurse,$opt_title,$opt_verbose);
> +
> +    unshift @ARGV, split ' ', $Config{pod2html} if $Config{pod2html};
> 
> I would at least localize |@ARGV|, either with |local|, or even:
> 

I think this is worthy of consideration.  rjbs earlier had some concerns 
about @ARGV.  But on the other hand the point of this project is not to 
cure all of Pod-Html's problems.  It's to get its test suite and 
internal skeleton in shape so that we can deal with those other problems.

> |my @_argv = @ARGV ; unshift @_argv, split( ' ', $Config{pod2html} || 
> ''); my $result = GetOptionsFromArray( \@_argv, 'backlink!' => 
> \$opt_backlink, 'cachedir=s' => \$opt_cachedir, 'css=s' => \$opt_css, 
> 'flush' => \$opt_flush, 'help' => \$opt_help, 'header!' => \$opt_header, 
> 'htmldir=s' => \$opt_htmldir, 'htmlroot=s' => \$opt_htmlroot, 'index!' 
> => \$opt_index, 'infile=s' => \$opt_infile, 'outfile=s' => 
> \$opt_outfile, 'poderrors!' => \$opt_poderrors, 'podpath=s' => 
> \$opt_podpath, 'podroot=s' => \$opt_podroot, 'quiet!' => \$opt_quiet, 
> 'recurse!' => \$opt_recurse, 'title=s' => \$opt_title, 'verbose!' => 
> \$opt_verbose, ); |
> 
> I see some other opportunitys to organize this subroutine for clarity 
> and maintenace - would you be willing to accept a PR for it after it 
> gets an initial merge?
> 

Yes.

Thank you very much.
Jim Keenan

rjbs commented on ext/Pod-Html/lib/Pod/Html.pm:
```diff
@@ -229,206 +220,236 @@ This program is distributed under the Artistic License.
+    }
 
+    $globals->{Title} //= "";
+    return html_escape($globals->{Title});
```
If I'm reading this correctly, this computes an HTML-escaped title, but *stores* a non-escaped version.  But then I see that when it's *called* it does this:
```
    $globals->{Title} = set_Title_from_podtree($globals, $podtree);
```

…so I think there some confusion, here.  The method call is called "set title" and does, but it returns a different value than it sets, which is immediately used to re-set the value!  This seems a bit too convoluted for my tastes.

jkeenan replied:
> If I'm reading this correctly, this computes an HTML-escaped title, but _stores_ a non-escaped version. But then I see that when it's _called_ it does this:
> 
> ```
>     $globals->{Title} = set_Title_from_podtree($globals, $podtree);
> ```
> 
> …so I think there some confusion, here. The method call is called "set title" and does, but it returns a different value than it sets, which is immediately used to re-set the value! This seems a bit too convoluted for my tastes.

Hmm, we'll have to think about this a bit.  I tried adding debugging statements to sub `set_Title_from_podtree`:
```
diff --git a/ext/Pod-Html/lib/Pod/Html.pm b/ext/Pod-Html/lib/Pod/Html.pm
index 1bb6e93e54..13a548d74c 100644
--- a/ext/Pod-Html/lib/Pod/Html.pm
+++ b/ext/Pod-Html/lib/Pod/Html.pm
@@ -425,7 +425,10 @@ sub set_Title_from_podtree {
     }
 
     $globals->{Title} //= "";
-    return html_escape($globals->{Title});
+print STDERR "AAA: $globals->{Title}\n";
+    my $rv = html_escape($globals->{Title});
+print STDERR "BBB: $rv\n";
+    return $rv;
 }
```
However, when I ran the test suite, in every case the text printed following `AAA:` was matched exactly by that following `BBB:` -- which means there's not enough variation in the test suite for an empirical judgment.
```
../ext/Pod-Html/t/crossref3.t .. 
1..1
AAA: htmlcrossref - Test HTML cross reference links
BBB: htmlcrossref - Test HTML cross reference links
ok 1 - cross references
ok
../ext/Pod-Html/t/eol.t ........ 
1..3
AAA: eol
BBB: eol
AAA: eol
BBB: eol
AAA: eol
BBB: eol
ok 1 - CR vs LF
ok 2 - CR vs CRLF
ok 3 - LF vs CRLF
ok
...
```
More thought needed.

jkeenan replied:
> If I'm reading this correctly, this computes an HTML-escaped title, but _stores_ a non-escaped version. But then I see that when it's _called_ it does this:
> 
> ```
>     $globals->{Title} = set_Title_from_podtree($globals, $podtree);
> ```
> 
> …so I think there some confusion, here. The method call is called "set title" and does, but it returns a different value than it sets, which is immediately used to re-set the value! This seems a bit too convoluted for my tastes.

Yes, but to a certain extent this is a side effect of the fact that, for review purposes, I've broken this project into 5 branches and 5 pull requests, each one of which produces correctly functioning code but only the last of which we really want to see in blead.

Currently, in blead (Pod-Html v1.27), we set `$Title` like this:
```
261 sub init_globals {
...
289     $Title = undef;             # title to give the pod(s)
290 }
291 
292 sub pod2html {
293     local(@ARGV) = @_;
294     local $_;
295 
296     init_globals();
297     parse_command_line();
...
379     unless(defined $Title) {
380     if($podtree->[0] eq "Document" && ref($podtree->[2]) eq "ARRAY" &&
381         $podtree->[2]->[0] eq "head1" && @{$podtree->[2]} == 3 &&
382         ref($podtree->[2]->[2]) eq "" && $podtree->[2]->[2] eq "NAME" &&
383         ref($podtree->[3]) eq "ARRAY" && $podtree->[3]->[0] eq "Para" &&
384         @{$podtree->[3]} >= 3 &&
385         !(grep { ref($_) ne "" }
386             @{$podtree->[3]}[2..$#{$podtree->[3]}]) &&
387         (@$podtree == 4 ||
388             (ref($podtree->[4]) eq "ARRAY" &&
389             $podtree->[4]->[0] eq "head1"))) {
390         $Title = join("", @{$podtree->[3]}[2..$#{$podtree->[3]}]);
391     }
392     }
393 
394     $Title //= "";
395     $Title = html_escape($Title);
```
You can tell that the block from lines 379 to 395 is ripe for encapsulation into a subroutine.  And that was done, albeit in a manner you find sub-optimal, in the 2nd of the 5 pull requests, the one we're currently discussing, https://github.com/jkeenan/perl5/pull/4#, which corresponds to branch [jkeenan:pod-html-oo-20210511](https://github.com/jkeenan/perl5/pull/4#).

But if we jump ahead to the 5th of the 5 p.r.s/branches, I think this becomes clearer.  At https://github.com/jkeenan/perl5/blob/pod-html-oo-20210512-02/ext/Pod-Html/lib/Pod/Html.pm#L520, the subroutine has now become a method call:
```
520 sub set_Title_from_podtree {
521     my ($self, $podtree) = @_;
522     unless(defined $self->{Title}) {
523         if($podtree->[0] eq "Document" && ref($podtree->[2]) eq "ARRAY" &&
524             $podtree->[2]->[0] eq "head1" && @{$podtree->[2]} == 3 &&
525             ref($podtree->[2]->[2]) eq "" && $podtree->[2]->[2] eq "NAME" &&
526             ref($podtree->[3]) eq "ARRAY" && $podtree->[3]->[0] eq "Para" &&
527             @{$podtree->[3]} >= 3 &&
528             !(grep { ref($_) ne "" }
529                 @{$podtree->[3]}[2..$#{$podtree->[3]}]) &&
530             (@$podtree == 4 ||
531                 (ref($podtree->[4]) eq "ARRAY" &&
532                 $podtree->[4]->[0] eq "head1"))) {
533             $self->{Title} = join("", @{$podtree->[3]}[2..$#{$podtree->[3]}]);
534         }
535     }
536 
537     $self->{Title} //= "";
538     return $self;
539 }
```
... whose syntax is very clean:
```
245     $self->set_Title_from_podtree($podtree);
```
So what you see in this p.r. is an *intermediate step* where certain edges are deliberately rough and await sanding in the 3 subsequent p.r.s.

However, as I've indicated, one of my reasons for requesting a preliminary review of this project in this repository is to facilitate review once this gets submitted to Perl5.git. Hence, if you think revising `set_Title_from_podtree` *here* will enable this p.r. to be digested more smoothly *there*, I'm willing to do so.  Perhaps something like this:
```
    $globals->{Title} //= "";
    $globals->{Title} = html_escape($globals->Title});
    return $globals;
}
```
Let me know what you think.

rjbs reviewed (changes requested):
I do think the nit I picked should be addressed.  Other than that, I feel pretty confident that I reviewed a previous version of this change, and so won't be attempting too close a read of this revision.

Leont committed:
Remove :win32 PerlIO layer

khwilliamson committed:
Pod::Simple: synch with CPAN version 3.43

jkeenan commented:
This p.r. is now superseded by the 2nd of 5 pull requests into Perl 5 blead: https://github.com/Perl/perl5/pull/18950.

Thank you very much.
JIm Keenan
